Refactor Events

jmjatlanta opened the pull request:
> **Note:** This is a re-do of pr #440, but based on the dev branch.

I would like to use this PR to open a dialog into integrating heavier use of C++ features into the codebase. Comments are encouraged. But please send any comments about what I should or should not be working on directly to me instead of commenting here.

Status as of 2021-08-10: I believe the code sufficiently demonstrates concepts that could be used in further refactoring. I believe the code to be bug-free. ~~, but not complete, as both old and new ideologies coexist for discussion and testing purposes.~~ More test cases are needed for edge cases that I have as yet not learned about.

Points for discussion:
1. Implementations in header files: Much of the code I have seen contains little implementation in .cpp files and most in .h header files. The disadvantages of this approach are (1) increased compile times and (2) delicate dependencies in linking libraries or executables. Moving the implementations to .cpp files does have disadvantages as well, such as objects compiled with different parameters can cause different (sometimes hard to diagnose) errors, but it is my opinion that the benefits outweigh the dangers.
2. Use the standard library where it makes sense to avoid platform-specific code. Two changes highlight the benefits here. Mutexes (std::mutex insted of pthread's version), and lists (std::list vs ptr-to-ptr lists).
3. Remove use of pointers where possible (i.e. pass references if possible, const references are even better), use reference counted pointers instead of raw pointers. The std::list vs ptr-to-ptr is an example here as well.
4. Declare necessary variables when you need them, not at the beginning of the function. This often makes the code easier to read.
5. Use exceptions where it makes sense. If the code is littered with `if` statements for error checking, consider using try/catch. This can reduce the code you need to write, as well as provide a performance boost in some instances if done correctly.
6. Use namespaces. Instead of starting each method name with `komdo_`, consider using namespace komodo. This reduces name collisions within the codebase (we know what an "event" is in komdo, and don't have to worry about puting "komodo_" in front of it, or a UI event having a name collision with it). This also often makes using your code as a library less cumbersome.
7. Use bool if you only need true/false instead of other integrals. This makes your intent clear.
8. Comment things that may be unclear. I personally use doxygen style comments on top of my functions. I hope everyone can understand my code by reading it, but sometimes it is very inconvenient to do so. Of course the evil side of comments is they are often stale or inaccurate. But the intent often remains.
9. Push implementations to where they belong. A "notary event" needs to be serialized. It is probably best that the code to do so reside close to the definition of what a notary event is. Every serialized "event" has a header, so this is probably a good time to use a hierarchy of "notary_event is an event, and knows how to serialize itself". Now reading or writing it to disk only requires code to tell a notary event to serialize/deserialize itself. The implementation code stays centralized and out of the way.
10. Refactor a little at a time. Knowing then what is known now, everyone would love to rewrite pretty much everything we've ever written. Avoid the urge. Bite-sized subsystems can be tackled. Eating the whole elephant will take time. The majority of the `komodo_events.h` file breaks point 9 above. I'd love to get rid of it, as well as big chunks of `komodo.h` for the same reason. But that would make testing and comparing the existing implementation with the refactor too much to take on in this round. So they will remain until they are no longer needed.
11. Header files should be independent. Include all the `#include` files necessary for the `.h`. This makes chasing down dependencies less of a problem at the expense of a little compile time. Use `#pragma once` to help with this.

The above points are simply opinions. I am not here to start wars. If the team decides that any (or all) of these are a bad idea, I'll happily work without using them. Explaining why they are a bad idea helps everyone, but is not required.

My apologies in advance to the reviewers. There is a lot of code movement with (relatively speaking) minor changes. Moving implementations from .h to .cpp touched many seemingly unrelated areas. Doing that in a separate PR would have made both reviews easier, but merging more complicated.

Reading suggestions and References: 
- [Migrating Legacy Systems](https://www.amazon.com/Migrating-Legacy-Systems-Interfaces-Incremental/dp/1558603301/ref=sr_1_2?dchild=1&keywords=migrating+legacy+systems&qid=1620656648&sr=8-2)
- [Komodo Contributing Guide](https://github.com/KomodoPlatform/komodo/blob/master/CONTRIBUTING.md) (includes recommending established standards such as [Google's](https://google.github.io/styleguide/cppguide.html) or [Bjarne Stroustrup](http://www.stroustrup.com/bs_faq2.html))
- [Bitcoin Contributor Readme](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md)
- [Bitcoin Developer Notes and Coding Style Guidelines](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md)

jmjatlanta commented:
Comment moved from #440 

> do we need both memread and the new mem_read functions?

I do not think so. I left both as I build out tests, and once I verify all cases are covered, we can eliminate the unnecessary functions.

DeckerSU commented:
@jmjatlanta I could be wrong, but looks like `komodo_state::add_event` adding an event in `std::list<std::shared_ptr<komodo::event>> events` [here](https://github.com/KomodoPlatform/komodo/pull/476/files#diff-10bcb1d8edc16456d7b98fedb3834fc3adf81d945d59963596beba41781530b4R30). according to initial logic [komodo_event_rewind](https://github.com/KomodoPlatform/komodo/pull/476/files#diff-6f60bd65b0d978eea08ff72a4dca9189d57a218d57d60f09e4bc6edf49443977R128) should do the opposite operation - remove event from list in case of height of event >= passed height in `komodo_event_rewind` and call `komodo_event_undo`. but looks like refactored [komodo_event_rewind](https://github.com/KomodoPlatform/komodo/pull/476/files#diff-6f60bd65b0d978eea08ff72a4dca9189d57a218d57d60f09e4bc6edf49443977R128-R150) still relies on `Komodo_events / Komodo_numevents` struct members, which is not used anymore. I guess [this loop](https://github.com/KomodoPlatform/komodo/pull/476/files#diff-6f60bd65b0d978eea08ff72a4dca9189d57a218d57d60f09e4bc6edf49443977R139-R148) should be refactored also. For example, to something like:
```
    while ( sp->events.size() > 0 )
    {
        std::shared_ptr<komodo::event> event = events.back();
        if ( event->height < height ) /* this should work as the base class for all events have height field */
            break;
        komodo_event_undo(sp,events.get()); /* passing raw ptr here should be safe, as we just read height inside */
        events.pop_back();
    }
```

DeckerSU commented:
Also i created this repository https://github.com/DeckerSU/events-refactor-test in which i combined old and refactored events implementations into separate source `events-tests.cpp` with some stubs, etc., to make tests "easier" and "lightweight". Old implementation is in `events_old` namespace and your one (new) is in `events_new` namespace. On a test file `komodostate` from real chain we have following results:
```
komodostate: reading 1772KB

read_count = 95309
rewind_count = 917
sp->NUM_NPOINTS = 8043
sp->Komodo_numevents = 92006

read_count = 95309
rewind_count = 917
sp->NUM_NPOINTS = 8043
sp->Komodo_numevents = 0
sp->events.size() = 95309
```
As we can see old implementation have `92006` events after parsing and new `95309` events due to bug described above in `komodo_event_rewind` implementation.

DeckerSU commented:
Also looks like in case `func == 'R'` (`komodo::event_opreturn`) logic is bit broken. In original implementation (non-refactored) it reads `txid`, `v`, `ovalue`, `olen`, `opret` variables from file / memory and inside `komodo_eventadd_opreturn` **ONLY in case** `if ( ASSETCHAINS_SYMBOL[0] != 0 )` fill the values, adds an event (1) and execute `komodo_opreturn` (2):
```
    void komodo_eventadd_opreturn(struct komodo_state *sp,char *symbol,int32_t height,uint256 txid,uint64_t value,uint16_t vout,uint8_t *buf,uint16_t opretlen)
    {
        struct komodo_event_opreturn O; uint8_t *opret;
        if ( ASSETCHAINS_SYMBOL[0] != 0 )
        {
            opret = (uint8_t *)calloc(1,sizeof(O) + opretlen + 16);
            O.txid = txid;
            O.value = value;
            O.vout = vout;
            memcpy(opret,&O,sizeof(O));
            memcpy(&opret[sizeof(O)],buf,opretlen);
            O.oplen = (int32_t)(opretlen + sizeof(O));
            komodo_eventadd(sp,height,symbol,KOMODO_EVENT_OPRETURN,opret,O.oplen); // (1) only if ASSETCHAINS_SYMBOL[0] != 0
            free(opret);
            if ( sp != 0 )
                komodo_opreturn(height,value,buf,opretlen,txid,vout,symbol); // (2) only if ASSETCHAINS_SYMBOL[0] != 0
        }
    }
```
New, refactored implementation haven't such `ASSETCHAINS_SYMBOL[0]` check:
```
void komodo_eventadd_opreturn( komodo_state *sp, char *symbol, int32_t height, std::shared_ptr<komodo::event_opreturn> opret)
{
    if ( sp != nullptr )
    {
        sp->add_event(symbol, height, opret); // (1)
        komodo_opreturn(height, opret->value, opret->opret.data(), opret->opret.size(), opret->txid, opret->vout, symbol); // (2)
    }
}
```
Of course, we have `ASSETCHAINS_SYMBOL[0]` check inside (1), but (2) will execute anyway, but shouldn't. My offer is to add condition check in `komodo_eventadd_opreturn` as in original implementation:
```
void komodo_eventadd_opreturn( komodo_state *sp, char *symbol, int32_t height, std::shared_ptr<komodo::event_opreturn> opret)
    {
        if ( sp != nullptr )
        {
            if ( ASSETCHAINS_SYMBOL[0] != 0 ) {
                sp->add_event(symbol, height, opret);
                komodo_opreturn(height, opret->value, opret->opret.data(), opret->opret.size(), opret->txid, opret->vout, symbol);
            }
        }
    }
```

DeckerSU commented:
**p.s.** added fixes in my `events-refactor-test` in `fixtry` branch:
1. for `komodo_event_rewind` - https://github.com/DeckerSU/events-refactor-test/blob/fixtry/events-tests.cpp#L923-L935
2. for `komodo_eventadd_opreturn` - https://github.com/DeckerSU/events-refactor-test/blob/fixtry/events-tests.cpp#L945-L954
so, you can have a look. now seems current `old` and `new` tests passed:
```
read_count = 95309
rewind_count = 917
sp->NUM_NPOINTS = 8043
sp->Komodo_numevents = 92006

read_count = 95309
rewind_count = 917
sp->NUM_NPOINTS = 8043
sp->Komodo_numevents = 0
sp->events.size() = 92006

0b6154c662432d4a6ba9ba9ee72388b576da4069d5e5aa0f33f0a8952395c895 - 0b6154c662432d4a6ba9ba9ee72388b576da4069d5e5aa0f33f0a8952395c895
a5fa719ad82e881342c9bae94c85e3d85cef2808e8bb3cad4d8b37a62afc9624 - a5fa719ad82e881342c9bae94c85e3d85cef2808e8bb3cad4d8b37a62afc9624
0000000000000000000000000000000000000000000000000000000000000000 - 0000000000000000000000000000000000000000000000000000000000000000
2524240 - 2524240
2524240 - 2524240
2524210 - 2524210
1628975442 - 1628975442
8043 - 8043
```
so, you could look.

jmjatlanta commented:
Thank you @DeckerSU for your detailed analysis and review. Fixes added. I will also attempt to add some testing around ASSETCHAINS_SYMBOL being filled in or not. A very nice catch.

dimxy commented:
could we improve struct notarized_checkpoint *NPOINTS in komodo_state
and komodo_state structure of KOMODO_STATES as well?
- I cannot see where NPOINTS is freed. Maybe we can have it as a std::vector or other std container?
- KOMODO_STATES maybe is good to hide from other src files (make it static) as it is accessed only via functions like komodo_stateptrget. Also its size is checked differently by an undefined constant (33 or 34, why this number?) Looks like it could be converted into std::map. Also komodo_stateptrget could return a reference instead of pointer (not to confuse the caller that is was malloc'ed) and this will remove limit of 33 elements. Plus the char *base param in komodo_stateptrget could be a const string&

DeckerSU commented on src/Makefile.am:
```diff
@@ -32,13 +32,13 @@ BITCOIN_INCLUDES += -I$(srcdir)/snark/libsnark
 if TARGET_WINDOWS
-LIBBITCOIN_SERVER=libbitcoin_server.a -lcurl
+LIBBITCOIN_SERVER=libbitcoin_server.a
 endif
 if TARGET_DARWIN
-LIBBITCOIN_SERVER=libbitcoin_server.a -lcurl
+LIBBITCOIN_SERVER=libbitcoin_server.a
 endif
 if TARGET_LINUX
-LIBBITCOIN_SERVER=libbitcoin_server.a -lcurl
+LIBBITCOIN_SERVER=libbitcoin_server.a
 endif
```
Minor improvement, as there is no difference between platforms now, we could change these:
```
if TARGET_WINDOWS
LIBBITCOIN_SERVER=libbitcoin_server.a
endif
if TARGET_DARWIN
LIBBITCOIN_SERVER=libbitcoin_server.a
endif
if TARGET_LINUX
LIBBITCOIN_SERVER=libbitcoin_server.a
endif
```
On just:
```
LIBBITCOIN_SERVER=libbitcoin_server.a
```

jmjatlanta replied:
Done

DeckerSU commented on src/komodo_structs.cpp:
```diff
@@ -0,0 +1,348 @@
+    uint16_t oplen;
+    if ( fread(&oplen,1,sizeof(oplen),fp) != sizeof(oplen) )
+        throw parse_error("Unable to parse length of opreturn record");
+    std::unique_ptr<uint8_t> b(new uint8_t[oplen]);
+    if ( fread(b.get(), 1, oplen, fp) != oplen)
+        throw parse_error("Unable to parse binary data of opreturn");
+    this->opret = std::vector<uint8_t>( b.get(), b.get() + oplen);
```
Is any reason why we need dynamic array allocation here, using std::unique_ptr and read `opret` content into intermediate variable `b`? I mean, as we have `oplen` known we could do something like direct reading in vector:

```
        uint16_t oplen;
        if ( fread(&oplen,1,sizeof(oplen),fp) != sizeof(oplen) )
            throw parse_error("Unable to parse length of opreturn record");
        this->opret.resize(oplen);
        if ( fread(&this->opret[0], sizeof this->opret[0], this->opret.size(), fp) != oplen )
            throw parse_error("Unable to parse binary data of opreturn");
```

jmjatlanta replied:
Good solution. Thanks.

DeckerSU commented:
> * Also its size is checked differently by an undefined constant (33 or 34, why this number?)

I guess size of:
https://github.com/KomodoPlatform/komodo/blob/db5b016cf7b8ffffe440ad774dee8c6815ce5726/src/komodo_globals.h#L43
It's correlated with:
https://github.com/KomodoPlatform/komodo/blob/db5b016cf7b8ffffe440ad774dee8c6815ce5726/src/komodo_globals.h#L134-L137

Just `CURRENCIES` should be declared as `CURRENCIES[MAX_CURRENCIES][8]` and `KOMODO_STATES` should be declared as `struct komodo_state KOMODO_STATES[MAX_CURRENCIES+2]`. 
```
int32_t komodo_baseid(char *origbase)
{
    int32_t i; char base[64];
    for (i=0; origbase[i]!=0&&i<sizeof(base); i++)
        base[i] = toupper((int32_t)(origbase[i] & 0xff));
    base[i] = 0;
    for (i=0; i<=MAX_CURRENCIES; i++)
        if ( strcmp(CURRENCIES[i],base) == 0 )
            return(i);
    //printf("illegal base.(%s) %s\n",origbase,base);
    return(-1);
}
...
struct komodo_state *komodo_stateptrget(char *base)
{
    int32_t baseid;
    if ( base == 0 || base[0] == 0 || strcmp(base,(char *)"KMD") == 0 )
        return(&KOMODO_STATES[33]);
    else if ( (baseid= komodo_baseid(base)) >= 0 )
        return(&KOMODO_STATES[baseid+1]);
    else return(&KOMODO_STATES[0]);
}
```
Why `+2` ? Bcz `komodo_baseid("KMD")` returns index `32` for KMD. And `komodo_stateptrget("KMD")` returns the pointer on the next element in `KOMODO_STATES` structure, i.e. element with index `33`. So, to fit the element with index 33 `KOMODO_STATES` should have size of `34` elements.

So, all seems to be correct, just need to bind these numbers (33, 34) to `MAX_CURRENCIES` constant everywhere to clear everything.

Example:
```
    std::cout << komodo_baseid("KMD") << std::endl;
    std::cout << (komodo_stateptrget("KMD")-&KOMODO_STATES[0]) << std::endl;
```
Output:
```
32
33
```

jmjatlanta commented:
> could we improve struct notarized_checkpoint *NPOINTS in komodo_state
> and komodo_state structure of KOMODO_STATES as well?
> 
>     * I cannot see where NPOINTS is freed. Maybe we can have it as a std::vector or other std container?
> 
>     * KOMODO_STATES maybe is good to hide from other src files (make it static) as it is accessed only via functions like komodo_stateptrget. Also its size is checked differently by an undefined constant (33 or 34, why this number?) Looks like it could be converted into std::map. Also komodo_stateptrget could return a reference instead of pointer (not to confuse the caller that is was malloc'ed) and this will remove limit of 33 elements. Plus the char *base param in komodo_stateptrget could be a const string&

As for refactoring `NPOINTS`, yes. It has very little to do with events, but I can do it as part of this PR if you wish. I may not get rid of `last_NPOINTSi` as I am unsure of exactly what it is doing. Perhaps I will be able to figure it out during the refactor.

As for hiding KOMODO_STATES from the layers above, I am all for it. But making it a static I would not. With a bit of thought, I'm sure we can find a place to store the collection that makes sense.

DeckerSU commented:
My offer may be will be bit strange from first look - but, i would like to offer don't touch anything else in this PR. Why? It's simple. Every PR should be "finalized" at certain point and contains only fixes / additions related to it's initial sense. I mean this PR is for _refactoring events_, but we already removed marmara stuff in it, moved some routines from .h (headers) to .cpp files and did some other things. Let's leave `NPOINTS` refactoring for the "second stage" / other PR. Bcz if we didn't do it - it could turn out in "infinite refactoring". Like ... what if i will say that storing events in plain binary file `komodostate` and having indexes `komodostate.ind` is suboptimal? :yum: We could use _LevelDB_ or even _SQLite3_ DB for that, which is perfectly fits in data / indexes conception and probably have additional mechanisms for data guard, in compare, with just binary file with records? Just want to offer stop adding changes in this PR and focus on double checking everything instead. What will you say?

jmjatlanta commented:
> My offer may be will be bit strange from first look - but, i would like to offer don't touch anything else in this PR. Why? It's simple. Every PR should be "finalized" at certain point and contains only fixes / additions related to it's initial sense. I mean this PR is for _refactoring events_, but we already removed marmara stuff in it, moved some routines from .h (headers) to .cpp files and did some other things. Let's leave `NPOINTS` refactoring for the "second stage" / other PR. Bcz if we didn't do it - it could turn out in "infinite refactoring". Like ... what if i will say that storing events in plain binary file `komodostate` and having indexes `komodostate.ind` is suboptimal? yum We could use _LevelDB_ or even _SQLite3_ DB for that, which is perfectly fits in data / indexes conception and probably have additional mechanisms for data guard, in compare, with just binary file with records? Just want to offer stop adding changes in this PR and focus on double checking everything instead. What will you say?

Strange? By no means. "Scope Creep" is a persistent problem, and I agree with your comment. Unfortunately, I committed it before your reminder was read. Consider my wrist slapped. We'll let @dimxy decide if he wants me to put it in another branch.

DeckerSU commented on src/komodo_notary.cpp:
```diff
@@ -299,38 +297,42 @@ struct notarized_checkpoint *komodo_npptr_for_height(int32_t height, int *idx)
-                return(np->notarized_height);
+            if ( pt.MoM != zero )
+                return pt.notarized_height;
         }
```
It seems wrong order of `NPOINTS[i]` scan here. It was _from end to begin_ with exit when `MoM` is not zero, but you did it vice-versa order - _from begin to end_. I guess better to revert range-loop changes here to something like:
```
for (int32_t i=sp->NPOINTS.size()-1; i>=0; i--)
{
    if ( sp->NPOINTS[i].MoM != zero )
        return sp->NPOINTS[i].notarized_height;
}
```
Or use reverse iterators ... or range-loop with `boost::adaptors::reverse`.

DeckerSU replied:
**p.s.** I vote for revert https://github.com/KomodoPlatform/komodo/pull/476/commits/de6166bddd3ab27e501b28dba5bbbe795c3251c7 from this PR.

jmjatlanta replied:
Agree. I attempted to knock it out too quickly. I'll move that to another branch. That'll give me more time to see what I can do about that last_NPOINTSi thing.

dimxy commented:
No problem, we could put aside NPOINTS and KOMODO_STATES refactoring (but I believe we need to do this one time and add comments as it is not easy to get the idea of those objects but we have to deal with them in code pretty often. Also it seems NPOINTS is never freed)

dimxy commented:
https://github.com/KomodoPlatform/komodo/pull/476#issuecomment-899707880
this looks a bit weird and better to be refactored.
What would happen for example if someone makes a chain with the name as one of cryptocurrencies in the list?

jmjatlanta commented:
> [#476 (comment)](https://github.com/KomodoPlatform/komodo/pull/476#issuecomment-899707880)
> this looks a bit weird and better to be refactored.
> What would happen for example if someone makes a chain with the name as one of cryptocurrencies in the list?

Agreed. The link between CURRENCY and komodo_state is pretty weak. I'm thinking the collection of komodo_state should be dynamic. My first reaction is a map keyed on the symbol, but I need to see all the use cases. I believe this and NPOINTS should become github issues.

dimxy commented:
> > [#476 (comment)](https://github.com/KomodoPlatform/komodo/pull/476#issuecomment-899707880)
> > this looks a bit weird and better to be refactored.
> > What would happen for example if someone makes a chain with the name as one of cryptocurrencies in the list?
> 
> Agreed. The link between CURRENCY and komodo_state is pretty weak. I'm thinking the collection of komodo_state should be dynamic. My first reaction is a map keyed on the symbol, but I need to see all the use cases. I believe this and NPOINTS should become github issues.
Yes a map looks like a great option and improves readability.
Also good to document the idea of the komodo_state and currency list here, seems this is a part of some app in the code (pax). And, there was earlier a suggestion to move some code into 'experimental' branches and clean the production branches, maybe this code part related to currencies may be just sent there

DeckerSU commented:
This line:
https://github.com/KomodoPlatform/komodo/blob/56590a4d1af683e2496feddf47f3df345dda04fb/src/komodo_bitcoind.h#L35
Could be removed as well, as we don't have anything Marmara related anymore.

As well as:
```
src/cc/marmara.cpp
```
Also don't needed now.

**p.s.** To clarify what i doing now - as you probably know we have also [Komodo-Qt (KomodoOcean)](https://github.com/DeckerSU/KomodoOcean) codebase which is almost similar to original komodod, but with GUI (Qt) part and various improvements. As codebases bit different this PR changes couldn't be easily cherry-picked or merged, so, i transfer changes manually "step-by-step", verifying and checking everything once again. From one side this is additional work and "time waste", but from other side - this is additional level of checking and testing which wouldn't be superflous in our case, so, don't be surprised if i'll mention this PR in Komodo-Qt changes.

dimxy commented:
> Could be removed as well, as we don't have anything Marmara related anymore.
all marmara remains need to be removed starting with ASSETCHAINS_MARMARA var (marmara now is a dedicated repo)
there is some if-else logic for marmara in staking code in komodo_bitcoind so let's be accurate with that (I did this in the research branch plz check it if any questions)

DeckerSU commented:
@jmjatlanta Seems found a serious bug, that lead to daemon crash during connecting a block + writing an event to `komodostate` file. Already digged a root of issue:
```
struct event_notarized : public event
{
    event_notarized() : event(komodo_event_type::EVENT_NOTARIZED, 0), notarizedheight(0), MoMdepth(0) {}
    event_notarized(int32_t ht) : event(EVENT_NOTARIZED, ht), notarizedheight(0), MoMdepth(0) {}
    event_notarized(uint8_t* data, long &pos, long data_len, int32_t height, bool includeMoM = false);
    event_notarized(FILE* fp, int32_t ht, bool includeMoM = false);
    uint256 blockhash;
    uint256 desttxid;
    uint256 MoM; 
    int32_t notarizedheight;
    int32_t MoMdepth; 
    char dest[16];
};
```
Let's look on `dest` field, it have size of 16, as it was in original. Ok, let's remember it. Now let's look inside `komodo_stateupdate`:
```
        else if ( height != 0 )
        {
            if ( sp != 0 )
            {
                std::shared_ptr<komodo::event_notarized> evt = std::make_shared<komodo::event_notarized>(height);
                memcpy(evt->dest, dest, 65); // (!!!)
                evt->blockhash = sp->NOTARIZED_HASH;
                evt->desttxid = sp->NOTARIZED_DESTTXID;
                evt->notarizedheight = sp->NOTARIZED_HEIGHT;
                evt->MoM = sp->MoM;
                evt->MoMdepth = sp->MoMdepth;
                write_event(evt, fp);
                komodo_eventadd_notarized(sp,symbol,height,evt);
            }
        }
```
What we doing in case of `event_notarized`? Copying 65 (!) symbols to the field which have only 16 bytes size? And on next step inside  `write_event` and execution of operator `<<` we have an unhandled exception. 

```
#0  0x00007ffff60d6438 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff60d803a in __GI_abort () at abort.c:89
#2  0x00007ffff61187fa in __libc_message (do_abort=do_abort@entry=2, fmt=fmt@entry=0x7ffff6231fd8 "*** Error in `%s': %s: 0x%s ***\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff612138a in malloc_printerr (ar_ptr=<optimized out>, ptr=<optimized out>, str=0x7ffff6232078 "free(): corrupted unsorted chunks", action=3) at malloc.c:5020
#4  _int_free (av=<optimized out>, p=<optimized out>, have_lock=0) at malloc.c:3874
#5  0x00007ffff612558c in __GI___libc_free (mem=<optimized out>) at malloc.c:2975
#6  0x0000555555a28708 in __gnu_cxx::new_allocator<char>::deallocate (this=0x7fff2fff5e20, __p=<optimized out>) at /usr/include/c++/5/ext/new_allocator.h:110
#7  std::allocator_traits<std::allocator<char> >::deallocate (__a=..., __n=<optimized out>, __p=<optimized out>) at /usr/include/c++/5/bits/alloc_traits.h:517
#8  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy (__size=<optimized out>, this=0x7fff2fff5e20) at /usr/include/c++/5/bits/basic_string.h:185
#9  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose (this=0x7fff2fff5e20) at /usr/include/c++/5/bits/basic_string.h:180
#10 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string (this=0x7fff2fff5e20, __in_chrg=<optimized out>) at /usr/include/c++/5/bits/basic_string.h:543
#11 std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::~basic_stringbuf (this=0x7fff2fff5dd8, __in_chrg=<optimized out>) at /usr/include/c++/5/sstream:65
#12 std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >::~basic_stringstream (this=0x7fff2fff5dc0, __in_chrg=<optimized out>, __vtt_parm=<optimized out>) at /usr/include/c++/5/sstream:718
#13 write_event (evt=std::shared_ptr (count 2, weak 0) 0x7fff1c127c60, fp=0x55567341cfc0) at komodo.cpp:186
#14 0x0000555555a2a8e5 in komodo_stateupdate (height=2499613, notarypubs=notarypubs@entry=0x0, numnotaries=numnotaries@entry=0 '\000', notaryid=notaryid@entry=0 '\000', txhash=..., voutmask=voutmask@entry=0, numvouts=numvouts@entry=0 '\000', pvals=0x0, numpvals=0 '\000', KMDheight=0, KMDtimestamp=0, opretvalue=0, opretbuf=0x0, opretlen=0, vout=0, MoM=..., MoMdepth=40) at komodo.cpp:302
#15 0x0000555555a2b766 in komodo_voutupdate (fJustCheck=<optimized out>, isratificationp=<optimized out>, notaryid=notaryid@entry=-1, scriptbuf=0x7fff2fff8980 "jH\257\360\345\064\006\rN'\257$\337\025\020\023ݳ\330\376N\322B\357\257\036\340\tFhY2\225\n\020$&", scriptlen=<optimized out>, height=2499613, txhash=..., i=7, j=1, voutmaskp=0x7fff2fff6c10, specialtxp=0x7fff2fff6bf0, notarizedheightp=0x7fff2fff6bf4, value=0, notarized=1, signedmask=716577440, timestamp=1627488174) at komodo.cpp:516
#16 0x0000555555a2c9e8 in komodo_connectblock (fJustCheck=fJustCheck@entry=false, pindex=<optimized out>, block=...) at komodo.cpp:728
#17 0x00005555556a2c99 in ConnectBlock (block=..., state=..., pindex=pindex@entry=0x7fff1c03da50, view=..., fJustCheck=fJustCheck@entry=false, fCheckPOW=fCheckPOW@entry=true) at main.cpp:3906
#18 0x00005555556aaa41 in ConnectTip (pblock=0x7fff2fffdca0, pindexNew=0x7fff1c03da50, state=...) at main.cpp:4296
...
```

Why earlier (before refactoring) we didn't faced with such error? Answer is in old `komodo_eventadd_notarized`:
```
struct komodo_event_notarized N;
...
strncpy(N.dest,dest,sizeof(N.dest)-1);
```
It was `strncpy` with sizeof(N.dest)-1m but in refactored implementation we trying to put 65 symbols in `event_notarized->dest`.

I fixed this error in Komodo-Qt (KomodoOcean) codebase the following way:

https://github.com/DeckerSU/KomodoOcean/commit/5f99d88568e0a49ce654135d59cb8e191f5667cd

In this case we can't change `dest` field definition in `event_notarized` to `char dest[KOMODO_ASSETCHAIN_MAXLEN];` and do write like ` memcpy(evt->dest, dest, KOMODO_ASSETCHAIN_MAXLEN);`, bcz existing `komodostate` files already have records and assuming size of dest field as `16`. That's why i leave structure and dest field untouched.

jmjatlanta commented:
> Seems found a serious bug, that lead to daemon crash during connecting a block + writing an event to `komodostate` file. Already digged a root of issue:

Yes, this is serious. Thank you for the forensic analysis. I am attempting to determine why I put a 65 there, and it may be because of thinking about the true size of `KOMODO_ASSETCHAIN_MAXLEN`, although why wouldn't I have used that instead of hard-coding 65?

In any case, you are correct in saying that for compatibility reasons, that field must remain 16 bytes, and we are limited to reading 15 of them. I will adjust, test, and look around for similar issues with `dest`.

DeckerSU commented on src/Makefile.am:
```diff
@@ -91,6 +83,12 @@ $(LIBUNIVALUE): $(wildcard univalue/lib/*)
 $(LIBCRYPTOCONDITIONS): $(wildcard cryptoconditions/src/*) $(wildcard cryptoconditions/include/*)
 	$(AM_V_at)$(MAKE) $(AM_MAKEFLAGS) -C $(@D) $(@F) OPTFLAGS="-O2 -march=x86-64 -g "
 
+%.o: %.c
+	$(CC) -c -o $@ $<
+
+$(LIBCJSON): cJSON.o komodo_cJSON.o komodo_cutils.o
+	$(AR) cr $(LIBCJSON) $^
+
```
btw, about the way how `libcjson.a` builds, may be the following approach is more clearly:
https://github.com/DeckerSU/KomodoOcean/blob/5f99d88568e0a49ce654135d59cb8e191f5667cd/src/Makefile.am#L77-L82
```
# libcjson build
LIBCJSON=libcjson.a
libcjson_a_SOURCES = cJSON.c \
  komodo_cJSON.c \
  komodo_cutils.cpp
libcjson_a_CPPFLAGS=-fPIC
EXTRA_LIBRARIES += $(LIBCJSON)
```
it's just like "open discussion" how it would better.

**p.s.** other tests of this PR still continues on my side.

DeckerSU replied:
Also need to add clean of `libcjson.a` on `make clean`, like this:
https://github.com/DeckerSU/KomodoOcean/commit/53ddb012484223747e1ef18140864209d32e8311

DeckerSU commented:
@jmjatlanta Windows build of your `github.com/jmjatlanta/komodo` `jmj_dev_event_3` failed with:
```
In file included from komodo_jumblr.h:16,
                 from komodo_jumblr.cpp:15:
uthash.h:895:5: error: ‘uint32_t’ does not name a type; did you mean ‘wint_t’?
     uint32_t count;
     ^~~~~~~~
     wint_t
uthash.h:909:5: error: ‘uint32_t’ does not name a type; did you mean ‘wint_t’?
     uint32_t expand_mult;
     ^~~~~~~~
     wint_t
uthash.h:919:5: error: ‘uint32_t’ does not name a type; did you mean ‘wint_t’?
     uint32_t num_buckets, log2_num_buckets;
     ^~~~~~~~
     wint_t
...
n file included from komodo_jumblr.h:20,
                 from komodo_jumblr.cpp:15:
/usr/share/mingw-w64/include/wincrypt.h:20:11: error: ‘LONG’ does not name a type; did you mean ‘LHND’?
   typedef LONG HRESULT;
           ^~~~
           LHND
In file included from komodo_jumblr.h:20,
                 from komodo_jumblr.cpp:15:
/usr/share/mingw-w64/include/wincrypt.h:238:11: error: ‘ULONG_PTR’ does not name a type; did you mean ‘MAXLONG_PTR’?
   typedef ULONG_PTR HCRYPTHASH;
           ^~~~~~~~~
           MAXLONG_PTR
/usr/share/mingw-w64/include/wincrypt.h:239:11: error: ‘ULONG_PTR’ does not name a type; did you mean ‘MAXLONG_PTR’?
   typedef ULONG_PTR HCRYPTKEY;
           ^~~~~~~~~
           MAXLONG_PTR
...
<tons of other similar errors>
...
make: *** [Makefile:3385: libbitcoin_common_a-komodo_jumblr.o] Error 1
```
I already made a fix for that:
https://github.com/DeckerSU/KomodoOcean/commit/058fa15e878c3dddad05fd7d03c14e1af91a278f

Please, push the similar commit into `jmj_dev_event_3` to fix the error and to add fix in this PR.

DeckerSU commented:
@jmjatlanta Other error with build `test-komodo/test_events.cpp` for WIndows:
```
test-komodo/test_events.cpp: In member function ‘virtual void TestEvents::TestEvents_komodo_faststateinit_test_Test::TestBody()’:
test-komodo/test_events.cpp:172:61: error: no match for ‘operator+’ (operand types are ‘const string_type’ {aka ‘const std::__cxx11::basic_string<wchar_t>’} and ‘const char [12]’)
             const std::string temp_filename = temp.native() + "/kstate.tmp";
                                               ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
In file included from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/stl_algobase.h:67,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/char_traits.h:39,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/ios:40,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/ostream:38,
                 from /home/decker/jmj_dev_event_3/depends/x86_64-w64-mingw32/share/../include/gtest/gtest.h:55,
                 from test-komodo/test_events.cpp:1:
/usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/stl_iterator.h:400:5: note: candidate: ‘template<class _Iterator> std::reverse_iterator<_Iterator> std::operator+(typename std::reverse_iterator<_Iterator>::difference_type, const std::reverse_iterator<_Iterator>&)’
     operator+(typename reverse_iterator<_Iterator>::difference_type __n,
     ^~~~~~~~
/usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/stl_iterator.h:400:5: note:   template argument deduction/substitution failed:
test-komodo/test_events.cpp:172:63: note:   mismatched types ‘const std::reverse_iterator<_Iterator>’ and ‘const char [12]’
             const std::string temp_filename = temp.native() + "/kstate.tmp";
                                                               ^~~~~~~~~~~~~
In file included from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/stl_algobase.h:67,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/char_traits.h:39,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/ios:40,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/ostream:38,
                 from /home/decker/jmj_dev_event_3/depends/x86_64-w64-mingw32/share/../include/gtest/gtest.h:55,
                 from test-komodo/test_events.cpp:1:
/usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/stl_iterator.h:1195:5: note: candidate: ‘template<class _Iterator> std::move_iterator<_IteratorL> std::operator+(typename std::move_iterator<_IteratorL>::difference_type, const std::move_iterator<_IteratorL>&)’
     operator+(typename move_iterator<_Iterator>::difference_type __n,
     ^~~~~~~~
/usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/stl_iterator.h:1195:5: note:   template argument deduction/substitution failed:
test-komodo/test_events.cpp:172:63: note:   mismatched types ‘const std::move_iterator<_IteratorL>’ and ‘const char [12]’
             const std::string temp_filename = temp.native() + "/kstate.tmp";
                                                               ^~~~~~~~~~~~~
In file included from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/string:52,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/locale_classes.h:40,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/ios_base.h:41,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/ios:42,
                 from /usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/ostream:38,
                 from /home/decker/jmj_dev_event_3/depends/x86_64-w64-mingw32/share/../include/gtest/gtest.h:55,
                 from test-komodo/test_events.cpp:1:
/usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/basic_string.h:5927:5: note: candidate: ‘template<class _CharT, class _Traits, class _Alloc> std::__cxx11::basic_string<_CharT, _Traits, _Alloc> std::operator+(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&, const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&)’
     operator+(const basic_string<_CharT, _Traits, _Alloc>& __lhs,
     ^~~~~~~~
/usr/lib/gcc/x86_64-w64-mingw32/8.3-posix/include/c++/bits/basic_string.h:5927:5: note:   template argument deduction/substitution failed:
test-komodo/test_events.cpp:172:63: note:   mismatched types ‘const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>’ and ‘const char [12]’
             const std::string temp_filename = temp.native() + "/kstate.tmp";
                                                               ^~~~~~~~~~~~~
```
This i didn't fix bcz Komodo-Qt have disabled tests build under Windows, but good to fix it anyway in `komodo` repo.

DeckerSU commented:
> > @jmjatlanta Other error with build `test-komodo/test_events.cpp` for WIndows:
> 
> Thank you. I currently do not have a working Windows build environment (WIP). I will update this with the status of my testing when (if?) it is done (seems gmp doesn't compile on WSL/Ubuntu 20.04 right now).

I used `Debian GNU/Linux 10 (buster)` with `mingw-w64/stable 6.0.0-3` (`gcc version 8.3-posix 20190406 (GCC)`) without any issues, except mentioned above. Probably will test building under Ubuntu 20.04 and WSL/Ubuntu 20.04 later, after i'll finished my current tests. Anyway, now we are closer to accept this #476 .

DeckerSU reviewed (approved):
LGTM now.

ca333 reviewed (approved):
secure code reviewed
